### ISSUE TITLE
Downloads are sometimes incomplete

### ISSUE THREAD

**Issue by Tobi823** [Tobi823/ffupdater#358]
I moved your error message to this issue @PsykoRider

<p dir="auto">Stacktrace:</p>
<div class="snippet-clipboard-content notranslate position-relative overflow-auto"><pre class="notranslate"><code class="notranslate">de.marmaro.krt.ffupdater.installer.exceptions.InstallationFailedException: Failed to install FIREFOX_NIGHTLY in the background with SESSION_INSTALLER.
	at de.marmaro.krt.ffupdater.BackgroundJob.installApplication(BackgroundJob.kt:266)
	at de.marmaro.krt.ffupdater.BackgroundJob.access$installApplication(BackgroundJob.kt:46)
	at de.marmaro.krt.ffupdater.BackgroundJob$installApplication$1.invokeSuspend(BackgroundJob.kt:0)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:104)
	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:570)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:749)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:677)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:664)
Caused by: de.marmaro.krt.ffupdater.installer.exceptions.InstallationFailedException: Can't validate the signature of the APK file.
	at de.marmaro.krt.ffupdater.installer.impl.AbstractAppInstaller.install2Internal(AbstractAppInstaller.kt:32)
	at de.marmaro.krt.ffupdater.installer.impl.AbstractAppInstaller.access$install2Internal(AbstractAppInstaller.kt:15)
	at de.marmaro.krt.ffupdater.installer.impl.AbstractAppInstaller$startInstallation$2.invokeSuspend(AbstractAppInstaller.kt:21)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
	at kotlinx.coroutines.internal.LimitedDispatcher.run(LimitedDispatcher.kt:42)
	at kotlinx.coroutines.scheduling.TaskImpl.run(Tasks.kt:95)
	... 4 more
Caused by: java.lang.IllegalArgumentException: Can't extract the signature from APK file '/storage/emulated/0/Android/data/de.marmaro.krt.ffupdater/files/Download/org_mozilla_fenix_2023-03-01 04:01.apk', length: 42224939, absolutePath: /storage/emulated/0/Android/data/de.marmaro.krt.ffupdater/files/Download/org_mozilla_fenix_2023-03-01 04:01.apk, isFile: true
	at de.marmaro.krt.ffupdater.security.PackageManagerUtil.getPackageArchiveInfo(PackageManagerUtil.kt:48)
	at de.marmaro.krt.ffupdater.security.FingerprintValidator$checkApkFile$2.invokeSuspend(FingerprintValidator.kt:29)
	at de.marmaro.krt.ffupdater.security.FingerprintValidator$checkApkFile$2.invoke(FingerprintValidator.kt:0)
	at de.marmaro.krt.ffupdater.security.FingerprintValidator$checkApkFile$2.invoke(FingerprintValidator.kt:0)
	at kotlinx.coroutines.intrinsics.UndispatchedKt.startUndispatchedOrReturn(Undispatched.kt:89)
	at kotlinx.coroutines.BuildersKt__Builders_commonKt.withContext(Builders.common.kt:166)
	at kotlinx.coroutines.BuildersKt.withContext(Unknown Source:1)
	at de.marmaro.krt.ffupdater.security.FingerprintValidator.checkApkFile(FingerprintValidator.kt:28)
	at de.marmaro.krt.ffupdater.installer.impl.AbstractAppInstaller.install2Internal(AbstractAppInstaller.kt:29)
	... 10 more
</code></pre><div class="zeroclipboard-container position-absolute right-0 top-0">
    
      
    

    
  </div></div>
<p dir="auto">Device information:</p>

Key | Value
-- | --
FFUpdater version | 77.7.9 (135) release
Device | RMX3561 (RMX3561, RED8BEL1, k6895v1_64)
Manufacturer | realme (realme)
Supported ABIs | arm64-v8a, armeabi-v7a, armeabi
Android version | 13 (SDK: 33)
OS | dg02-pool03-kvm41, root, release-keys, 1672839940000

**Comment by Tobi823** [comment]
The downloaded file is only **42**'224'939 bytes long and not **91**'899'067 (arm64-v8a) or **85**'877'977 (armeabi-v7a).

https://firefox-ci-tc.services.mozilla.com/tasks/index/mobile.v3.firefox-android.apks.fenix-nightly.2023.03.01.revision.5373a91aac9411400eae2995f5185bb857879f63/arm64-v8a

**Comment by Tobi823** [comment]
So somehow the download was aborted and FFUpdater did not notice it.

**Comment by PsykoRider** [comment]
Stacktrace:
```
java.lang.RuntimeException: Failed to install FIREFOX_NIGHTLY in the foreground.
	at de.marmaro.krt.ffupdater.DownloadActivity.installApp(DownloadActivity.kt:404)
	at de.marmaro.krt.ffupdater.DownloadActivity.access$installApp(DownloadActivity.kt:59)
	at de.marmaro.krt.ffupdater.DownloadActivity$installApp$1.invokeSuspend(DownloadActivity.kt:0)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:104)
	at android.os.Handler.handleCallback(Handler.java:938)
	at android.os.Handler.dispatchMessage(Handler.java:99)
	at android.os.Looper.loopOnce(Looper.java:210)
	at android.os.Looper.loop(Looper.java:299)
	at android.app.ActivityThread.main(ActivityThread.java:8293)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:556)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1045)
Caused by: de.marmaro.krt.ffupdater.installer.exceptions.InstallationFailedException: Can't validate the signature of the APK file.
	at de.marmaro.krt.ffupdater.installer.impl.AbstractAppInstaller.install2Internal(AbstractAppInstaller.kt:32)
	at de.marmaro.krt.ffupdater.installer.impl.AbstractAppInstaller.access$install2Internal(AbstractAppInstaller.kt:15)
	at de.marmaro.krt.ffupdater.installer.impl.AbstractAppInstaller$startInstallation$2.invokeSuspend(AbstractAppInstaller.kt:21)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
	at kotlinx.coroutines.internal.LimitedDispatcher.run(LimitedDispatcher.kt:42)
	at kotlinx.coroutines.scheduling.TaskImpl.run(Tasks.kt:95)
	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:570)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:749)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:677)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:664)
Caused by: java.lang.IllegalArgumentException: Can't extract the signature from APK file '/storage/emulated/0/Android/data/de.marmaro.krt.ffupdater/files/Download/org_mozilla_fenix_2023-03-01 16:02.apk', length: 121965, absolutePath: /storage/emulated/0/Android/data/de.marmaro.krt.ffupdater/files/Download/org_mozilla_fenix_2023-03-01 16:02.apk, isFile: true
	at de.marmaro.krt.ffupdater.security.PackageManagerUtil.getPackageArchiveInfo(PackageManagerUtil.kt:48)
	at de.marmaro.krt.ffupdater.security.FingerprintValidator$checkApkFile$2.invokeSuspend(FingerprintValidator.kt:29)
	at de.marmaro.krt.ffupdater.security.FingerprintValidator$checkApkFile$2.invoke(FingerprintValidator.kt:0)
	at de.marmaro.krt.ffupdater.security.FingerprintValidator$checkApkFile$2.invoke(FingerprintValidator.kt:0)
	at kotlinx.coroutines.intrinsics.UndispatchedKt.startUndispatchedOrReturn(Undispatched.kt:89)
	at kotlinx.coroutines.BuildersKt__Builders_commonKt.withContext(Builders.common.kt:166)
	at kotlinx.coroutines.BuildersKt.withContext(Unknown Source:1)
	at de.marmaro.krt.ffupdater.security.FingerprintValidator.checkApkFile(FingerprintValidator.kt:28)
	at de.marmaro.krt.ffupdater.installer.impl.AbstractAppInstaller.install2Internal(AbstractAppInstaller.kt:29)
	... 10 more
```
Device information:
| Key | Value |
| --- | --- |
| FFUpdater version | 77.7.9 (135) release |
| Device | POCO F2 Pro (lmi_eea, lmi, lmi) |
| Manufacturer | POCO (Xiaomi) | 
| Supported ABIs | arm64-v8a, armeabi-v7a, armeabi |
| Android version | 12 (SDK: 31) |
| OS | non-pangu-pod-xztq7, builder, release-keys, 1672051495000 |

**Comment by PsykoRider** [comment]
Stacktrace:
```
de.marmaro.krt.ffupdater.FFUpdaterException: Downloaded or extracted APK file is not a valid ZIP file.
	at de.marmaro.krt.ffupdater.installer.ApkChecker$Companion.throwIfApkFileIsNoValidZipFile(ApkChecker.kt:24)
	at de.marmaro.krt.ffupdater.installer.ApkChecker$Companion$throwIfApkFileIsNoValidZipFile$1.invokeSuspend(ApkChecker.kt:0)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:104)
	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:570)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:749)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:677)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:664)
Caused by: java.util.zip.ZipException: error in opening zip file
	at java.util.zip.ZipFile.open(Native Method)
	at java.util.zip.ZipFile.<init>(ZipFile.java:265)
	at java.util.zip.ZipFile.<init>(ZipFile.java:187)
	at java.util.zip.ZipFile.<init>(ZipFile.java:201)
	at de.marmaro.krt.ffupdater.installer.ApkChecker$Companion$throwIfApkFileIsNoValidZipFile$2.invokeSuspend(ApkChecker.kt:21)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
	at kotlinx.coroutines.internal.LimitedDispatcher.run(LimitedDispatcher.kt:42)
	at kotlinx.coroutines.scheduling.TaskImpl.run(Tasks.kt:95)
	... 4 more
```
Device information:
| Key | Value |
| --- | --- |
| FFUpdater version | 77.7.11 (137) release |
| Device | POCO F2 Pro (lmi_eea, lmi, lmi) |
| Manufacturer | POCO (Xiaomi) | 
| Supported ABIs | arm64-v8a, armeabi-v7a, armeabi |
| Android version | 12 (SDK: 31) |
| OS | non-pangu-pod-xztq7, builder, release-keys, 1672051495000 |

**Comment by ledufff** [comment]
Stacktrace:
```
java.lang.RuntimeException: Failed to install MULL_FROM_REPO in the foreground.
	at de.marmaro.krt.ffupdater.DownloadActivity.installApp(DownloadActivity.kt:391)
	at de.marmaro.krt.ffupdater.DownloadActivity.access$installApp(DownloadActivity.kt:61)
	at de.marmaro.krt.ffupdater.DownloadActivity$installApp$1.invokeSuspend(DownloadActivity.kt:0)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:104)
	at android.os.Handler.handleCallback(Handler.java:942)
	at android.os.Handler.dispatchMessage(Handler.java:99)
	at android.os.Looper.loopOnce(Looper.java:211)
	at android.os.Looper.loop(Looper.java:300)
	at android.app.ActivityThread.main(ActivityThread.java:8143)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:580)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1028)
Caused by: de.marmaro.krt.ffupdater.installer.exceptions.InstallationFailedException: INSTALL_FAILED_ABORTED: Permission denied. The installation failed because it was actively aborted.
	at de.marmaro.krt.ffupdater.installer.impl.SessionInstaller.fail(SessionInstaller.kt:191)
	at de.marmaro.krt.ffupdater.installer.impl.SessionInstaller.handleAppInstallationResult(SessionInstaller.kt:127)
	at de.marmaro.krt.ffupdater.installer.impl.SessionInstaller.access$handleAppInstallationResult(SessionInstaller.kt:30)
	at de.marmaro.krt.ffupdater.installer.impl.SessionInstaller$registerIntentReceiver$1.onReceive(SessionInstaller.kt:211)
	at android.app.LoadedApk$ReceiverDispatcher$Args.lambda$getRunnable$0$android-app-LoadedApk$ReceiverDispatcher$Args(LoadedApk.java:1838)
	at android.app.LoadedApk$ReceiverDispatcher$Args$$ExternalSyntheticLambda0.run(Unknown Source:2)
	... 8 more
```
Device information:
| Key | Value |
| --- | --- |
| FFUpdater version | 77.7.14 (140) release |
| Device | 21081111RG (amber_global, amber, amber) |
| Manufacturer | Xiaomi (Xiaomi) | 
| Supported ABIs | arm64-v8a, armeabi-v7a, armeabi |
| Android version | 13 (SDK: 33) |
| OS | pangu-build-component-system-30644-kkkdm-rmqw0-rn3ml, builder, release-keys, 1672846836000 |

**Comment by PsykoRider** [comment]
Stacktrace:
```
de.marmaro.krt.ffupdater.network.exceptions.NetworkException: File was not downloaded: /storage/emulated/0/Android/data/de.marmaro.krt.ffupdater/files/Download/com_kiwibrowser_browser_4698988734.apk
	at de.marmaro.krt.ffupdater.installer.ApkChecker$Companion.throwIfDownloadedFileHasDifferentSize(ApkChecker.kt:29)
	at de.marmaro.krt.ffupdater.DownloadActivity.startDownload(DownloadActivity.kt:315)
	at de.marmaro.krt.ffupdater.DownloadActivity.access$startDownload(DownloadActivity.kt:61)
	at de.marmaro.krt.ffupdater.DownloadActivity$startDownload$1.invokeSuspend(DownloadActivity.kt:0)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
	at android.os.Handler.handleCallback(Handler.java:938)
	at android.os.Handler.dispatchMessage(Handler.java:99)
	at android.os.Looper.loopOnce(Looper.java:210)
	at android.os.Looper.loop(Looper.java:299)
	at android.app.ActivityThread.main(ActivityThread.java:8293)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:556)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1045)
```
Device information:
| Key | Value |
| --- | --- |
| FFUpdater version | 77.7.14 (140) release |
| Device | POCO F2 Pro (lmi_eea, lmi, lmi) |
| Manufacturer | POCO (Xiaomi) | 
| Supported ABIs | arm64-v8a, armeabi-v7a, armeabi |
| Android version | 12 (SDK: 31) |
| OS | non-pangu-pod-c2wbh, builder, release-keys, 1677763033000 |

**Comment by PsykoRider** [comment]
Stacktrace:
```
java.lang.RuntimeException: Failed to install KIWI in the foreground.
	at de.marmaro.krt.ffupdater.DownloadActivity.installApp(DownloadActivity.kt:391)
	at de.marmaro.krt.ffupdater.DownloadActivity.access$installApp(DownloadActivity.kt:61)
	at de.marmaro.krt.ffupdater.DownloadActivity$installApp$1.invokeSuspend(DownloadActivity.kt:0)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:104)
	at android.os.Handler.handleCallback(Handler.java:938)
	at android.os.Handler.dispatchMessage(Handler.java:99)
	at android.os.Looper.loopOnce(Looper.java:210)
	at android.os.Looper.loop(Looper.java:299)
	at android.app.ActivityThread.main(ActivityThread.java:8293)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:556)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1045)
Caused by: de.marmaro.krt.ffupdater.installer.exceptions.InstallationFailedException: INSTALL_FAILED_ABORTED: Permission denied. The installation failed because it was actively aborted.
	at de.marmaro.krt.ffupdater.installer.impl.SessionInstaller.fail(SessionInstaller.kt:191)
	at de.marmaro.krt.ffupdater.installer.impl.SessionInstaller.handleAppInstallationResult(SessionInstaller.kt:127)
	at de.marmaro.krt.ffupdater.installer.impl.SessionInstaller.access$handleAppInstallationResult(SessionInstaller.kt:30)
	at de.marmaro.krt.ffupdater.installer.impl.SessionInstaller$registerIntentReceiver$1.onReceive(SessionInstaller.kt:211)
	at android.app.LoadedApk$ReceiverDispatcher$Args.lambda$getRunnable$0$LoadedApk$ReceiverDispatcher$Args(LoadedApk.java:1694)
	at android.app.LoadedApk$ReceiverDispatcher$Args$$ExternalSyntheticLambda0.run(Unknown Source:2)
	... 8 more
```
Device information:
| Key | Value |
| --- | --- |
| FFUpdater version | 77.7.14 (140) release |
| Device | POCO F2 Pro (lmi_eea, lmi, lmi) |
| Manufacturer | POCO (Xiaomi) | 
| Supported ABIs | arm64-v8a, armeabi-v7a, armeabi |
| Android version | 12 (SDK: 31) |
| OS | non-pangu-pod-c2wbh, builder, release-keys, 1677763033000 |

**Comment by Tobi823** [comment]
I think this is fixed in 78.0.1. If not, please create a new issue.
